schnorr_fun/src/frost/shared_key.rs

```diff
@@ -0,0 +1,410 @@
+    }
+
+    /// The public key that has been shared.
+    ///
+    /// This is using *public key* in a rather loose sense. Unless it's a `SharedKey<EvenY>` then it
+    /// won't be usable as an actual Schnorr [BIP340] public key.
+    ///
+    /// [BIP340]: https://bips.xyz/340
+    pub fn public_key(&self) -> Point<T, Public, Z> {
+        // SAFETY: we hold the first coefficient to match the type parameters always
+        let public_key = Z::cast_point(self.point_polynomial[0]).expect("invariant");
+        T::cast_point(public_key).expect("invariant")
+    }
+
+    /// Encodes a `SharedKey` as the compressed encoding of each underlying polynomial coefficient
+    ///
+    /// i.e. call [`Point::to_bytes`] on each coefficent starting with the constant term. Note that
+    /// even if it's a `SharedKey<EvenY>` the first coefficient (A.K.A the public key) will still be
+    /// encoded as 33 bytes.
+    ///
+    /// ⚠ Unlike other secp256kfun things this doesn't exactly match the serde/bincode
+    /// implemenations which will length prefix the list of points.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes = Vec::with_capacity(self.point_polynomial.len() * 33);
+        for coeff in &self.point_polynomial {
+            bytes.extend(coeff.to_bytes())
+        }
+        bytes
+    }
+
+    /// Decodes a `SharedKey<T,Z>` (for any `T` and `Z`) from a slice.
+    ///
+    /// Returns `None` if the bytes don't represent points or if the first coefficient doesn't
+    /// satisfy the constraints of `T` and `Z`.
+    pub fn from_slice(bytes: &[u8]) -> Option<Self> {
+        let mut poly = vec![];
+        for point_bytes in bytes.chunks(33) {
+            poly.push(Point::from_slice(point_bytes)?);
+        }
+
+        // check first coefficient satisfies both type parameters
+        let first_coeff = Z::cast_point(poly[0])?;
+        let _check = T::cast_point(first_coeff)?;
+
+        Some(Self::from_inner(poly))
+    }
+}
+
+impl SharedKey<Normal> {
+    /// Convert the key into a BIP340 "x-only" SharedKey.
+    ///
+    /// This is the [BIP340] compatible version of the key which you can put in a segwitv1 output.
+    ///
+    /// [BIP340]: https://bips.xyz/340
+    pub fn into_xonly(mut self) -> SharedKey<EvenY> {
+        let needs_negation = !self.public_key().is_y_even();
+        if needs_negation {
+            self = self.homomorphic_negate();
+            debug_assert!(self.public_key().is_y_even());
+        }
+
+        SharedKey::from_inner(self.point_polynomial)
+    }
+}
+
+impl SharedKey<Normal, Zero> {
+    /// Constructor to create a shared key from a vector of points where each item represent a polynomial
+    /// coefficient.
+    ///
+    /// The resulting shared key will be `SharedKey<Normal, Zero>`. It's up to the caller to do the zero check with [`non_zero`]
+    ///
+    /// [`non_zero`]: Self::non_zero
+    pub fn from_poly(poly: Vec<Point<Normal, Public, Zero>>) -> Self {
```

Comment:
Is this again because there's nothing special about preventing `Zero` for a shared key?

I always feel weird seeing it allowed. Why avoid panics when the user passes an empty poly for the shared key? (maybe they just don't know it but have a secret share?)